sph/include/sph/kernel_ve/density_kern.hpp

```diff
@@ -69,20 +68,24 @@ CUDA_DEVICE_HOST_FUN inline void densityJLoop(int i, T sincIndex, T K, const cst
         T   vloc   = dist * hInv;
         T   w      = ::sphexa::math::pow(lt::wharmonic_lt_with_derivative(wh, whd, vloc), sincIndex);
         T   dw     = wharmonic_derivative(vloc, w) * sincIndex;
-        T   dterh  = -(3.0 * w + vloc * dw);
+        T   dterh  = -(T(3) * w + vloc * dw);
         T   xmassj = m[j] / rho0[j];
 
         kxi += w * xmassj;
         whomegai += dterh * xmassj;
     }
 
-    kx[i]    = K * (kxi + xmassi) * h3Inv;
-    whomegai = K * (whomegai - 3.0 * xmassi) * h3Inv * hInv;
+    kxi *= K * h3Inv;
 
-    ro[i]    = kx[i] * m[i] / xmassi;
-    whomegai = whomegai * m[i] / xmassi + (ro[i] / rho0i - K * xmassi * h3Inv) * wrho0i;
-    T dhdrho = -hi / ro[i] / 3.0; // This /3 is the dimension hard-coded.
-    gradh[i] = 1.0 - dhdrho * whomegai;
+    whomegai = K * (whomegai - T(3) * xmassi) * h3Inv * hInv;
```

Comment:
If the self-contribution initialization for whomegai is done before the loop, you can change this to:
whomegai *= K * h3Inv * hInv

Reply:
nice simplification